octree_creation_app/params.py

```diff
@@ -27,69 +36,131 @@
+            refinements = collect_refinements_from_dict(values)
+            if refinements:
+                msg = (
+                    "Detected deprecated 'Refinement A property' style refinements,"
+                    " converting to a list of dictionaries."
+                )
+                warnings.warn(msg)
+            values["refinements"] = refinements
+
+        return values
+
+    @model_serializer(mode="wrap")
+    def distribute_refinements(self, handler, info):
+        """Convert refinements to a individual parameters."""
+        dump = handler(self, info)
+        refinements = dump.pop("refinements")
+        refinement_params = {}
+        for i, group in enumerate(refinements):
+            group_id = string.ascii_uppercase[i]
+            for param, value in group.items():
+                param_type = "object" if param == "refinement_object" else param
+                param_name = f"Refinement {group_id} {param_type}"
+                refinement_params[param_name] = value
+
+        return dict(dump, **refinement_params)
+
+    @classmethod
+    def build(cls, input_data: InputFile | dict) -> Self:
+        """
+        Build a dataclass from a dictionary or InputFile.
+
+        :param input_data: Dictionary of parameters and values.
+
+        :return: Dataclass of application parameters.
+        """
+
+        data = input_data
+
+        if isinstance(input_data, InputFile) and input_data.data is not None:
+            data = input_data.data.copy()
+
+        if not isinstance(data, dict):
+            raise TypeError("Input data must be a dictionary or InputFile.")
+
+        kwargs = OctreeParams.collect_input_from_dict(cls, data)  # type: ignore
+        out = cls(**kwargs)
+        if isinstance(input_data, InputFile):
+            out._input_file = input_data
+
+        return out
+
+    @staticmethod
+    def collect_input_from_dict(base_model: type[BaseModel], data: dict[str, Any]):
         """
-        Update parameters with dictionary contents.
+        Recursively replace BaseModel objects with dictionary of 'data' values.
 
-        :param params_dict: Dictionary of parameters.
+        :param base_model: BaseModel object holding data and possibly other nested
+            BaseModel objects.
+        :param data: Dictionary of parameters and values without nesting structure.
         """
+        update = {}
+        for field, info in base_model.model_fields.items():
+            if (
+                isinstance(info.annotation, type)
+                and not isinstance(info.annotation, GenericAlias)
+                and issubclass(info.annotation, BaseModel)
+            ):
+                update[field] = OctreeParams.collect_input_from_dict(
+                    info.annotation,
+                    data,  # type: ignore
+                )
+            else:
+                if field in data:
+                    update[field] = data.get(field, info.default)
```

Comment:
```suggestion
        update = super().collect_input_from_dict(base_mode, data)
```
Looks like mostly duplicate from the base class

Reply:
yep, good catch!  Fixed.